ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -547,6 +560,80 @@ let domainify_uname ~domain uname =
   else
     Printf.sprintf "%s@%s" uname domain
 
+let range s e step =
```

Comment:
All of this looks brittle to me. What are we trying to do here? We have an SQL database but seem to use low-level string scanning. What is encoded in this string that we are reading here?

Reply:
The `range` just like python `range`,  like `range 0 10 1 -> [0;1;2;3;4;5;6;7;8;9]`, see the unittest,

I suppose the question is about `parse_value_from_pbis`

This function parsing the values extracted from PBIS database (sqlite3), and decode it, see the unittest for the input and output.
The PBIS databases values are formatted as `X"xxxxxxxx"`,  where "xxxxxxx" are values in hex. 
All the data are formatted this way, like 
```
sqlite3 registry.db "select QUOTE(Value) from regvalues1 where ValueName='SamAccountName'"
X'4C004F00430041004C0048004F0053005400320024000000' 
```
For the parsing details, refer to the code, 

This migrate should only happens after update/upgrade from CH which joined domain with PBIS.

Reply:
This should be based around a function that takes a hex-string and returns the decoded string (a byte or char sequence). An easy way to do this is using `Hex.of_string` from the `hex` package.

Reply:
`Hex.of_string` is used finally , refer to the code, 
but before that, there are some preparing work like strip the tailing and heading, remvoe the splitting and ending `00` , etc,